apps/src/lib/node/ledger/protocol/mod.rs

```diff
@@ -181,6 +212,40 @@ where
     }
 }
 
+const TX_ETH_BRIDGE_WASM_NAME: &str = "tx_eth_bridge";
+
+fn construct_tx_eth_bridge(wasm_dir: &Path) -> Result<Tx> {
+    let tx_data = vec![];
+    tracing::debug!(
+        bytes = tx_data.len(),
+        "serialized tx_data for state update transaction"
+    );
+    let tx_code = {
+        let checksums = crate::wasm_loader::Checksums::read_checksums(wasm_dir);
```

Comment:
NB: we `safe_exit` the process here if a `checksums.json` can't be found, so we can't write a test for an empty wasm_dir yet (as the test runner exits similar to https://github.com/anoma/namada/issues/104)

Reply:
if we panicked instead of exiting the process, we would be able to catch the panic... bah but at that point, it would be better to return some `Result` type from `wasm_loader::Checksums::read_checksums`

Reply:
I'm not sure how I feel about doing this checksums check every block. It's a lot of io and a validator who uses the wrong wasm should be considered byzantine. Is there a way to refactor this to do the check once at start up?

Reply:
Then e2e tests would cover the fail to start up when there is an empty / nonexistent wasm dir

Reply:
Yeah it's not great that it reads and hashes every block. Possible solutions:

- we could bundle the wasm into the binary at compile time using `include_bytes!` (along with a precalculated hardcoded hash to check against once at start up)
- instead of `wasm_dir` we could use a `preloaded_wasms` (`HashMap<String, Vec<u8>>`) that is constructed once at start up time. It would only contain `tx_eth_bridge.wasm` for now.

Reply:
I think we can have something similar to the `Checksums(pub HashMap<String, String>)` struct that keeps `.wasm` files it has read and checked in memory. Then we can refactor `Shell`/`AbcippShim` slightly to have one of these components and pass it down to `apply_tx`.

Reply:
Can't we just use the wasm cache? It gets initialized in the `Shell::new` method and we can dump the wasm blob in there.

Reply:
`tx_wasm_cache` lets us look up a compiled wasm module by wasm code. We could warm it up using `tx_wasm_cache.fetch_or_compile(code)` in `Shell::new` though we'll still need the wasm code itself during `apply_tx` to get the compiled wasm out of the cache.

Reply:
We could write a fetch method that would return an error if the wasm code isn't in there.  Also in the Cache code, it uses a hash of the code internally for lookups. So our new fetch method could just take a hash as input

Reply:
:+1: I'll try implementing this, we should either pass the hash down to `apply_tx` somehow or look up the compiled wasm by some preknown name e.g. `tx_wasm_cache.fetch_by_name("tx_eth_bridge")`